File: src/main/java/at/hrechny/predictionsbot/database/repository/AuditRepository.java
package at.hrechny.predictionsbot.database.repository;

import at.hrechny.predictionsbot.database.entity.AuditEntity;
import at.hrechny.predictionsbot.database.model.ApiProvider;
import jakarta.inject.Singleton;
import jakarta.persistence.EntityManager;
import jakarta.transaction.Transactional;
import java.time.Instant;
import java.util.Optional;
import java.util.UUID;

@Singleton
public class AuditRepository {

  private final EntityManager entityManager;

  public AuditRepository(EntityManager entityManager) {
    this.entityManager = entityManager;
  }

  @Transactional
  public AuditEntity save(AuditEntity entity) {
    return entityManager.merge(entity);
  }

  public Optional<AuditEntity> findById(UUID id) {
    return Optional.ofNullable(entityManager.find(AuditEntity.class, id));
  }

  public int countAllByApiProviderAndApiKeyAndRequestDateAfter(ApiProvider apiProvider, String apiKey, Instant date) {
    return entityManager
        .createQuery("""
            select count(a)
            from AuditEntity a
            where a.apiProvider = :apiProvider
              and a.apiKey = :apiKey
              and a.requestDate > :date
            """, Long.class)
        .setParameter("apiProvider", apiProvider)
        .setParameter("apiKey", apiKey)
        .setParameter("date", date)
        .getSingleResult()
        .intValue();
  }
}

File: src/main/kotlin/at/hrechny/predictionsbot/database/repository/AuditRepository.kt
package at.hrechny.predictionsbot.database.repository

import at.hrechny.predictionsbot.database.entity.AuditEntity
import jakarta.inject.Singleton
import jakarta.persistence.EntityManager
import io.micronaut.transaction.annotation.Transactional
import java.time.Instant
import java.util.Optional
import java.util.UUID

@Singleton
@Transactional
class AuditRepository(
    private val entityManager: EntityManager,
) {
    @Transactional
    fun save(entity: AuditEntity): AuditEntity = entityManager.merge(entity)

    fun findById(id: UUID): Optional<AuditEntity> = Optional.ofNullable(entityManager.find(AuditEntity::class.java, id))

    fun countAllByConnectorNameAndRequestDateAfter(
        connectorName: String,
        date: Instant,
    ): Int = entityManager
        .createQuery(
            """
            select count(a)
            from AuditEntity a
            where a.connectorName = :connectorName
              and a.requestDate > :date
            """.trimIndent(),
            Long::class.javaObjectType,
        )
        .setParameter("connectorName", connectorName)
        .setParameter("date", date)
        .singleResult
        .toInt()
}
